aten/src/ATen/cudnn/Descriptors.h

```diff
@@ -0,0 +1,178 @@
+#pragma once
+
+#include "Exceptions.h"
+
+#include "cudnn-wrapper.h"
+#include <ATen/Tensor.h>
+
+namespace at { namespace cudnn {
+
+// TODO: Add constructors for all of the descriptors
+
+inline int dataSize(cudnnDataType_t dataType)
+{
+  switch (dataType) {
+    case CUDNN_DATA_HALF: return 2;
+    case CUDNN_DATA_FLOAT: return 4;
+    default: return 8;
+  }
+}
+
+// The stride for a size-1 dimensions is not uniquely determined; in
+// fact, it can be anything you want, because the fact that the
+// tensor is size 1 at this dimension means that you will never actually
+// try advancing your pointer by this stride.
+//
+// However, CuDNN has a much more stringent requirement on strides:
+// if you are passing a contiguous input, it better be the case
+// that the stride for dim i is the product of the sizes of dims
+// i+1 to the end.  This stride is indeed uniquely determined.  This
+// function modifies 'stride' in place so this invariant holds.
+static inline void fixSizeOneDimStride(int dim, const int *size, int *stride) {
+  int64_t z = 1;
+  for(int d = dim-1; d >= 0; d--)
+  {
+    if (size[d] == 1) {
+      stride[d] = z;
+    } else {
+      z *= size[d];
+    }
+  }
+}
+
+// This is NON-OWNING on tensor data
+struct TensorDescriptor
+{
+  cudnnTensorDescriptor_t desc;
+  void* ptr;
```

Comment:
I'm not sure data ptr should be part of TensorDescriptor. In cudnn model (however foolish) the same descriptor can be shared (and is shared, e.g. in batch norm) between several tensors, that's why it only has strides and sizes info, on purpose.

Reply:
Hmm, this is an interesting point. I moved the data ptr into TensorDescriptor mostly to make the benchmark code look better after I deleted the `Convolution` struct, because otherwise I would have needed to add a big pile of extra parameters to thread the pointers through otherwise. I guess I can bring a `Convolution`-like struct to get rid of this situation.